src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -20,54 +20,46 @@
 #include <cstring>
 
 #include "cdrom.h"
-#include "channel_names.h"
 #include "support.h"
 
 #if defined(LINUX)
 #include <fcntl.h>
 #include <limits.h>
 #include <linux/cdrom.h>
+#include <mntent.h>
 #include <sys/ioctl.h>
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <unistd.h>
 
-// ioctl cannot be read more than 75 redbook frames at a time (one second of audio)
-constexpr int InputBufferMaxRedbookFrames = 25;
-constexpr size_t PcmSamplesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME / REDBOOK_BPS;
-constexpr size_t InputBufferMaxSamples = InputBufferMaxRedbookFrames *
-                                         PcmSamplesPerRedbookFrame;
-
 CDROM_Interface_Ioctl::~CDROM_Interface_Ioctl()
 {
-	if (mixer_channel) {
-		MIXER_DeregisterChannel(mixer_channel);
-	}
 	if (IsOpen()) {
 		close(cdrom_fd);
 	}
+	cdrom_fd = -1;
 }
 
 bool CDROM_Interface_Ioctl::IsOpen() const
 {
 	return cdrom_fd != -1;
 }
 
-bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, std::string& upc)
 {
 	if (!IsOpen()) {
 		return false;
 	}
 
-	struct cdrom_mcn cdrom_mcn;
-	int ret = ioctl(cdrom_fd, CDROM_GET_MCN, &cdrom_mcn);
+	struct cdrom_mcn cdrom_mcn = {};
 
-	if (ret > 0) {
+	if (ioctl(cdrom_fd, CDROM_GET_MCN, &cdrom_mcn) == 0) {
 		attr = 0;
-		safe_strncpy(upc, (char *)cdrom_mcn.medium_catalog_number, 14);
+		upc = (const char *)cdrom_mcn.medium_catalog_number;
```

Comment:
Note that this is not perfectly safe (it fails when `medium_catalog_number` is not zero-terminated).

Reply:
Using something like `safe_strncpy` seems safer, yeah.

Reply:
I would suggest adding the following functions to string_utils.h and use one of these:
```
template<size_t N>
std::string safe_tostring(char (&str)[N]) noexcept
{
	return std::string(str, safe_strlen(str));
}

std::string safe_tostring(const char* str, const std::size_t maxlen) noexcept
{
	return std::string(str, strnlen(str, maxlen));
}

```

Reply:
Sure, those look like some good helper functions.  Added and used.